firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/SequentialReference.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.appdistribution.impl;
+
+import androidx.annotation.VisibleForTesting;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.firebase.concurrent.FirebaseExecutors;
+import java.util.concurrent.Executor;
+
+/**
+ * A reference to an object that uses a sequential executor to ensure non-concurrent reads and
+ * writes while preventing lock contention.
+ */
+class SequentialReference<T> {
+
+  private final Executor sequentialExecutor;
+  private T value;
+
+  /** Get a {@link SequentialReference} that controls access uses the given sequential executor. */
+  static SequentialReference withSequentialExecutor(Executor sequentialExecutor) {
+    return new SequentialReference(sequentialExecutor);
+  }
+
+  /**
+   * Get a {@link SequentialReference} that controls access using its own sequential executor backed
+   * by the given base executor.
+   */
+  static SequentialReference withBaseExecutor(Executor baseExecutor) {
+    return new SequentialReference(FirebaseExecutors.newSequentialExecutor(baseExecutor));
+  }
+
+  private SequentialReference(Executor sequentialExecutor) {
+    this.sequentialExecutor = sequentialExecutor;
+  }
+
+  /**
+   * Sets the value, returning a {@link Task} will complete once it is set.
+   *
+   * <p>For convenience when chaining tasks together, the result of the returned task is the value.
+   */
+  Task<T> set(T newValue) {
+    TaskCompletionSource<T> taskCompletionSource = new TaskCompletionSource<>();
+    sequentialExecutor.execute(
+        () -> {
+          value = newValue;
+          taskCompletionSource.setResult(value);
+        });
+    return taskCompletionSource.getTask();
+  }
+
+  /**
+   * Gets a {@link Task} that will complete with the value.
+   *
+   * <p>Unless a direct executer is passed to {@link Task#addOnSuccessListener(Executor,
+   * OnSuccessListener)} or similar methods, be aware that the value may have changed by the time it
+   * is used.
```

Comment:
I wonder if it's actually a problem, i.e. when directExecutor is used here's the sequence of events:

0. `state` = `0`
1. `get()` called(returns `0`)
2. a concurrent `set(1)` is queued but waits for `1` to complete
3. `get()` continuation executes and returns the transformed task using `0`
4. task from 3 is returned to the caller
5. 2 updates the value to `1`
6. the caller observes results based on `0` when it is actually now equal `1`

Now let's what happens when a "real" executor is used:

0. `state` = `0`
1. `get()` called(returns `0`)
2. a concurrent `set(1)` and updates the state to `1`
3. `get()` continuation executes and returns the transformed task using `0`
4. task from 3 is returned to the caller
6. the caller observes results based on `0` when it is actually now equal `1`

So regardless of the sequence of events, the outcome is the same, the value may have changed by the time the caller observes the result, unless direct executor is used throughout the whole continuation change(which generally is not a good idea).

Point being: consider using lightweight throughout unless there is some internal invariant that I am missing and will get violated by this.

Reply:
Yes, good call.

I think there's another reason not to necessarily recommend the direct executor. I was able to reproduce some slightly different behavior: if I introduce an artificial delay in the execution of the initial `get()`, then only a direct executor ensures the entire chain off of the `get()` will execute with the correct value before the first `set(1)` happens. This is because any callbacks get called on the sequential executor's thread immediately after the read or write happens. I was able to repro this.

But this was an artificial test, and it gets complicated in the case where the `get()` succeeds immediately. When the task is already complete before the handlers are added, the callbacks end up being called immediately on the **calling** thread, not the sequential executor's thread. And in that case there's no practical difference between using a direct executor or a "real" executor.

So a direct executor really doesn't guarantee an order of execution. More reason not to recommend it IMO. I think we just need to understand when data might be old and whether that's a problem.